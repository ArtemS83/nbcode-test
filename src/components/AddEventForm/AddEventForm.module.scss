.Form {
  margin-left: auto;
  margin-right: auto;
  max-width: 480px;
  height: auto;
  padding: 15px;
  border: 1px solid rgb(31, 26, 34);
  border-radius: 4px;

  text-align: start;
  background-color: rgb(125, 231, 25);
  box-shadow: 2px 6px 8px -2px #0ab40a;
}

.label {
  display: flex;
  margin-bottom: 35px;
  flex-direction: column;
  font-size: 20px;
  font-weight: 500;
  input {
    margin-top: 15px;
    padding: 20px;
    // max-width: 220px;
    height: 30px;
    font-size: 18px;
    background-color: rgba(239, 235, 241, 0.883);
    border: 1px solid rgb(72, 28, 233);
    border-radius: 4px;
    outline: none;
    box-shadow: 2px 6px 8px -2px #a09c9c;
    cursor: pointer;
    transition: border 300ms ease-in-out;
    &:focus,
    &:hover {
      border: 1px solid rgba(224, 247, 18, 0.568);
    }
  }
}
.divBtn {
  display: flex;
  justify-content: space-between;
  align-items: center;
}
