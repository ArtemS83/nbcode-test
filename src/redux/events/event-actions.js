import { createAction } from '@reduxjs/toolkit';

export const addEvent = createAction('event/addEvent');
