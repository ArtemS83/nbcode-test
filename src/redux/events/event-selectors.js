export const getEvents = state => state.items;
