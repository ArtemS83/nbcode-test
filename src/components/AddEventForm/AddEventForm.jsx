import { useState } from 'react';
import { useDispatch } from 'react-redux';
import { v4 as uuid } from 'uuid';
import styles from './AddEventForm.module.scss';
import Button from '../Button';
import { addEvent } from 'redux/events/event-actions';

const AddEventForm = () => {
  const [firstName, setFirstName] = useState('');
  const [lastName, setLastName] = useState('');
  const [email, setEmail] = useState('');
  const [date, setDate] = useState('');

  const dispatch = useDispatch();

  const handleInputChange = ({ target }) => {
    const { value, name } = target;
    switch (name) {
      case 'firstName':
        setFirstName(value);
        return;
      case 'lastName':
        setLastName(value);
        return;
      case 'email':
        setEmail(value);
        return;
      case 'date':
        setDate(value);
        return;

      default:
        return;
    }
  };

  const resetForm = () => {
    setFirstName('');
    setLastName('');
    setEmail('');
    setDate('');
  };

  const handleAddEvent = e => {
    e.preventDefault();
    if (!firstName || !lastName || !email || !date) {
      alert(`Please fill in all fields! `);
      return;
    }
    const newEvent = {
      id: uuid(),
      firstName,
      lastName,
      email,
      date,
    };
    dispatch(addEvent(newEvent));
    resetForm();
  };

  return (
    <form className={styles.Form} onSubmit={handleAddEvent}>
      <label className={styles.label}>
        First name
        <input
          type="text"
          name="firstName"
          value={firstName}
          pattern="^[a-zA-Zа-яА-Я]+(([' -][a-zA-Zа-яА-Я ])?[a-zA-Zа-яА-Я]*)*$"
          title="The name can only consist of letters, apostrophes, dashes and spaces. For example Adrian, Jacob Mercer, Charles de Batz de Castelmore d'Artagnan, etc."
          required
          placeholder="Enter First name "
          onChange={handleInputChange}
        />
      </label>
      <label className={styles.label}>
        Last name
        <input
          type="text"
          name="lastName"
          value={lastName}
          pattern="^[a-zA-Zа-яА-Я]+(([' -][a-zA-Zа-яА-Я ])?[a-zA-Zа-яА-Я]*)*$"
          title="The name can only consist of letters, apostrophes, dashes and spaces. For example Adrian, Jacob Mercer, Charles de Batz de Castelmore d'Artagnan, etc."
          required
          placeholder="Enter Last name"
          onChange={handleInputChange}
        />
      </label>
      <label className={styles.label}>
        Email
        <input
          type="text"
          name="email"
          value={email}
          pattern="^[\w]{1,}[\w.+-]{0,}@[\w-]{2,}([.][a-zA-Z]{2,}|[.][\w-]{2,}[.][a-zA-Z]{2,})$"
          title="Enter Email"
          required
          placeholder="Enter Email"
          onChange={handleInputChange}
        />
      </label>
      <label className={styles.label}>
        Event date
        <input
          type="date"
          name="date"
          value={date}
          pattern="\d{4}-\d{2}-\d{2}"
          title="Enter Event date"
          required
          placeholder="Enter Event date"
          onChange={handleInputChange}
        />
      </label>
      <div className={styles.divBtn}>
        <Button title="Reset form" onClick={resetForm} />
        <Button title="Add Event" type="submit" />
      </div>
    </form>
  );
};

export default AddEventForm;
