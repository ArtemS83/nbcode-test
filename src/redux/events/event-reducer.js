import { createReducer } from '@reduxjs/toolkit';
import { addEvent } from './event-actions';

const initialState = [];

export default createReducer(initialState, {
  [addEvent]: (state, action) => [...state, action.payload],
});
