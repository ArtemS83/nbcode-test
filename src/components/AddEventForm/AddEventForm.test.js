import { render } from '@testing-library/react';
import reducer from 'redux/events/event-reducer';

import AddEventForm from '../Button';
import { addEvent } from 'redux/events/event-actions';

describe('AddEventForm', () => {
  it('AddEventForm should render properly', () => {
    const { container } = render(<AddEventForm />);
    expect(container).toMatchSnapshot();
  });

  it('addEvent action should return a new object', () => {
    const newEvent = {
      id: 'f4c7a726-711b-4a55-8c40-1ec53f8f740b',
      firstName: 'Artem',
      lastName: 'Sergieiev',
      email: 'artsergeev83@gmail.com',
      date: '2021-09-23',
    };
    expect(addEvent(newEvent)).toMatchObject({
      type: 'event/addEvent',
      payload: newEvent,
    });
  });

  it('event reducer', () => {
    const initialState = [];
    const action = {
      type: 'event/addEvent',
      payload: {
        id: 'f4c7a726-711b-4a55-8c40-1ec53f8f740b',
        firstName: 'Artem',
        lastName: 'Sergieiev',
        email: 'artsergeev83@gmail.com',
        date: '2021-09-23',
      },
    };
    expect(reducer(initialState, action)).toEqual([
      ...initialState,
      action.payload,
    ]);
  });
});
