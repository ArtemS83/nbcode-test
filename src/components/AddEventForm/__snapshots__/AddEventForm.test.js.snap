// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddEventForm AddEventForm should render properly 1`] = `
<div>
  <button
    class="buttonButton"
    data-testid="onClickFn"
    type="button"
  >
    Click me
  </button>
</div>
`;
